ports/esp32/esp32_sigmadelta.c

```diff
@@ -0,0 +1,160 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Mike Causer <mcauser@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "modmachine.h"
+#include "mphalport.h"
+#include "driver/sigmadelta.h"
+
+// Forward declaration
+extern const mp_obj_type_t esp32_sigmadelta_type;
+
+typedef struct _esp32_sigmadelta_obj_t {
+    mp_obj_base_t base;
+    uint8_t channel_id;
+    gpio_num_t pin;
+    int8_t duty;
+    uint8_t prescale;
+} esp32_sigmadelta_obj_t;
+
+// Defined in machine_time.c; simply added the error message
+// Fixme: Should use this updated error hadline more widely in the ESP32 port.
+//        At least update the method in machine_time.c.
+STATIC esp_err_t check_esp_err(esp_err_t code) {
+    if (code) {
+        mp_raise_msg(&mp_type_OSError, esp_err_to_name(code));
+    }
+
+    return code;
+}
+
+STATIC mp_obj_t esp32_sigmadelta_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_channel,   MP_ARG_REQUIRED |                  MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_pin,       MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_duty,                        MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_prescale,                    MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 80} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    mp_uint_t channel_id = args[0].u_int;
+    gpio_num_t pin_id = machine_pin_get_id(args[1].u_obj);
+    mp_int_t duty = args[2].u_int;
+    mp_uint_t prescale = args[3].u_int;
+
+    if (duty < -128 || duty > 127) {
+        mp_raise_ValueError("duty must be between -128 and 127");
+    }
+
+    if (prescale < 0 || prescale > 255) {
```

Comment:
`prescale` is an `mp_uint_t` (typedef to `unsigned long`); any point testing for <0?

Reply:
`duty` should be a `int8_t` (-128 ~ 127). Am I casting it to a `mp_int_t` correctly (line 65)?

`prescale` should be `uint8_t` (0 - 255). If it's being cast as `unsigned`, then no, as it's not possible to be < 0, so that could be removed.

I thought `mp_uint_t` was an `int`, not a `long`?
`typedef unsigned int mp_uint_t;`

Reply:
Ah, good point, `mp_uint_t` is port-dependent. Haven't quite tracked down what that pops out to for ESP32 but it seems to usually be an unsigned type...